hyperv: improve networking code

Maners opened the pull request:
* make VM XML network interface always use 
````xml
<interface type="bridge"><source bridge="GUID"/></interface>
````
for both define and dumpxml actions
* implement virConectListAllNetworks, virNetworkGetXMLDesc and other networking methods so that we can get lists of Hyper-V networks by getting network handles and then get network UUIDs (as names may be duplicated)

Maners committed:
hyperv: fix switch case fall-through issues.

as detected by GCC7.

Maners committed:
hyperv: move hypervDebugResponseXml to hyperv_wmi

so that it can be called from anywhere in the driver, not just "v2" API.

Maners committed:
hyperv: attach VM to network using bridge type.

that is make it work with:

<interface type="bridge">
    <source bridge="GUID" />
</interface>

aside from consitency it also resembles syntax used by esx driver.

Maners committed:
hyperv: implement virConnectListAllNetworks.

and cleanup the network driver a little bit to reduce code duplication.

Maners committed:
hyperv: implement virNetworkGetXMLDesc.

Basic implementation merely showing name and uuid. Also implemented
autostart/persitence API functions as those are queried with virsh
net-list so we need those to display info accurately.

SriRamanujam commented on src/hyperv/hyperv_network_api_v1.c:
```diff
@@ -264,6 +264,81 @@ hyperv1ConnectNumOfNetworks(virConnectPtr conn)
+
+    virUUIDFormat(network->uuid, uuid_string);
+
+    if (virAsprintf(&filter, "Name = \"%s\"", uuid_string) < 0 ||
```
I think you missed the actual call to `hyperv1VirtualSwitch` here.

Maners replied:
yep, fixed.

SriRamanujam commented on src/hyperv/hyperv_api_v1.c:
```diff
@@ -1663,13 +1671,11 @@ hyperv1DomainAttachSyntheticEthernetAdapter(virDomainPtr domain,
                 "Name=\"%s\",SystemCreationClassName=\"Msvm_VirtualSwitch\","
-                "SystemName=\"%s\"", computer->data->ElementName,
-                switchport_guid_string, net->data.network.name) < 0)
+                "SystemName=\"%s\"", hostname,
```
Is the hostname from Libvirt always guaranteed to be what Windows expects?

Maners replied:
the hostname is passed from hypervNDomainDefineXML and is result of hypervNGetHostSystem call which in turn is Msvm_ComputerSystem->data->ElementName so we had 2 lookups here instead of using what we already have given to us by the caller.

SriRamanujam commented on src/hyperv/hyperv_api_v1.c:
```diff
@@ -1594,20 +1594,28 @@ hyperv1DomainAttachSyntheticEthernetAdapter(virDomainPtr domain,
     invokeXmlParam *params = NULL;
     properties_t *NewResources = NULL;
 
+    /* get the vSwitch first so we can get data from it as needed */
```
Why is this necessary? The only thing you need from this is the GUID of the Msvm_VirtualSwitch object, which is already present in the XML (`net->data.bridge.brname`). You even use it in the error message on line 1658.

Maners replied:
yeah that's left over from the previous iteration when I had the code work with network name instead of guid - this was looking up the vSwitch by name and then using guild down the line.

SriRamanujam commented on src/hyperv/hyperv_api_v1.c:
```diff
@@ -1594,20 +1594,28 @@ hyperv1DomainAttachSyntheticEthernetAdapter(virDomainPtr domain,
      */
     virBufferAddLit(&query, MSVM_VIRTUALSWITCH_V1_WQL_SELECT);
-    virBufferAsprintf(&query, "where Name = \"%s\"", net->data.network.name);
+    virBufferAsprintf(&query, "where Name = \"%s\"", vSwitch->data->Name);
```
Should just use the GUID from the XML.

SriRamanujam commented on src/hyperv/hyperv_api_v1.c:
```diff
@@ -1663,13 +1671,11 @@ hyperv1DomainAttachSyntheticEthernetAdapter(virDomainPtr domain,
-                "SystemName=\"%s\"", computer->data->ElementName,
-                switchport_guid_string, net->data.network.name) < 0)
+                "SystemName=\"%s\"", hostname,
+                switchport_guid_string, vSwitch->data->Name) < 0)
```
Should just use the GUID from the XML.

SriRamanujam commented on src/hyperv/hyperv_api_v2.c:
```diff
@@ -1716,22 +1715,19 @@ hyperv2DomainAttachSyntheticEthernetAdapter(virDomainPtr domain,
                 "Msvm_VirtualEthernetSwitch.CreationClassName="
                 "\"Msvm_VirtualEthernetSwitch\",Name=\"%s\"",
-                computer->data->ElementName, net->data.network.name) < 0)
+                hostname, vSwitch->data->Name) < 0)
```
Is the hostname always guaranteed to be what Windows expects?

Maners replied:
yes,  it's pulled from Msvm_Computersystem->data->ElementName by the caller and passed to this one as hostname

SriRamanujam commented on src/hyperv/hyperv_network_api_v1.c:
```diff
@@ -71,104 +71,142 @@ hyperv1MsvmVirtualSwitchToNetwork(virConnectPtr conn,
-hyperv1ConnectListNetworks(virConnectPtr conn, char **const names, int maxnames)
+static int
+hyperv1GetVirtualSwitchList(hypervPrivate *priv,
+                            const char *filter,
```
You don't clean up this filter string anywhere. It should probably be cleaned up in this function.

Maners replied:
this string is stack allocated in all callers, so there's nothing to clean-up

SriRamanujam commented on src/hyperv/hyperv_network_api_v1.c:
```diff
@@ -71,104 +71,142 @@ hyperv1MsvmVirtualSwitchToNetwork(virConnectPtr conn,
+    if (hyperv1GetVirtualSwitchList(priv, NULL, &vSwitchList) < 0)
+        goto cleanup;
+
+    if (networks) {
```
This entire block seems oddly laid out. I would first allocate `nets` big enough to hold all the networks (you have this number from `hyperv1GetVirtualSwitchList`), then run through the loop below without the call to `VIR_RESIZE_N`.

Maners replied:
fixed.

SriRamanujam commented on src/hyperv/hyperv_network_api_v1.c:
```diff
@@ -71,104 +71,142 @@ hyperv1MsvmVirtualSwitchToNetwork(virConnectPtr conn,
+            continue;
+        }
+
+        if (VIR_RESIZE_N(nets, nnets, count, 2) < 0)
```
Allocate the networks all at once above, don't resize in here.

SriRamanujam commented on src/hyperv/hyperv_network_api_v1.c:
```diff
@@ -71,104 +71,142 @@ hyperv1MsvmVirtualSwitchToNetwork(virConnectPtr conn,
+    Msvm_VirtualSwitch_V1 *vSwitch = NULL;
+    virNetworkPtr network = NULL;
+    virNetworkPtr *nets = NULL;
+    int count = 0;
```
Minor nitpick: this should be `size_t`.

SriRamanujam commented on src/hyperv/hyperv_network_api_v2.c:
```diff
@@ -71,103 +71,141 @@ hyperv2MsvmVirtualSwitchToNetwork(virConnectPtr conn,
-hyperv2ConnectListNetworks(virConnectPtr conn, char **const names, int maxnames)
+static int
+hyperv2GetVirtualSwitchList(hypervPrivate *priv,
+                            const char *filter,
```
Free this variable in the cleanup.

Maners replied:
again, this is stack-allocated by the caller so once it gets out of scope it will cleanup

SriRamanujam commented on src/hyperv/hyperv_network_api_v2.c:
```diff
@@ -71,103 +71,141 @@ hyperv2MsvmVirtualSwitchToNetwork(virConnectPtr conn,
+    Msvm_VirtualEthernetSwitch_V2 *vSwitch = NULL;
+    virNetworkPtr network = NULL;
+    virNetworkPtr *nets = NULL;
+    int count = 0;
```
Minor nitpick: this should be `size_t`.

Maners replied:
return type of this function is int and it eventually gets assigned to ret

SriRamanujam commented on src/hyperv/hyperv_network_api_v2.c:
```diff
@@ -71,103 +71,141 @@ hyperv2MsvmVirtualSwitchToNetwork(virConnectPtr conn,
+    if (hyperv2GetVirtualSwitchList(priv, NULL, &vSwitchList) < 0)
+        goto cleanup;
+
+    if (networks) {
```
Why not just allocate the full list of networks in this block and skip the `VIR_RESIZE_N` call in the loop below?

SriRamanujam commented on src/hyperv/hyperv_network_api_v2.c:
```diff
@@ -71,103 +71,141 @@ hyperv2MsvmVirtualSwitchToNetwork(virConnectPtr conn,
-    result = count;
+    for (vSwitch = vSwitchList; vSwitch != NULL; vSwitch = vSwitch->next) {
+        /* just count the networks */
+        if (!networks) {
```
This block should probably be outside of the loop, since you know how many networks there are via `hyperv2GetVirtualSwitchList`.

SriRamanujam commented on src/hyperv/hyperv_network_api_v1.c:
```diff
@@ -71,104 +71,142 @@ hyperv1MsvmVirtualSwitchToNetwork(virConnectPtr conn,
-    result = count;
+    for (vSwitch = vSwitchList; vSwitch != NULL; vSwitch = vSwitch->next) {
+        /* just count the networks */
+        if (!networks) {
```
This block should probably outside of this loop, since you know how many networks there are via `hyperv1GetVirtualSwitchList`.

SriRamanujam commented on src/hyperv/hyperv_network_api_v2.c:
```diff
@@ -71,103 +71,141 @@ hyperv2MsvmVirtualSwitchToNetwork(virConnectPtr conn,
-    virBuffer query = VIR_BUFFER_INITIALIZER;
-    Msvm_VirtualEthernetSwitch_V2 *virtualSwitchList = NULL;
-    Msvm_VirtualEthernetSwitch_V2 *virtualSwitch = NULL;
+    int count = 0;
```
Minor nitpick: this should be `size_t`.

Maners replied:
same as above, its used as return value on success so it should match the function's return type which is int

SriRamanujam commented on src/hyperv/hyperv_network_api_v2.c:
```diff
@@ -191,57 +228,53 @@ hyperv2NetworkLookupByName(virConnectPtr conn, const char *name)
-                &virtualSwitch) < 0) {
-        goto cleanup;
-    }
     if (virtualSwitch == NULL) {
```
Combine this with the block above: 
```C
if (hyperv2GetVirtualSwitchList(priv, filter, &virtualSwitch) || virtualSwitch == NULL) {
    virReportError(VIR_ERR_NO_NETWORK, _("No network found with name %s"), name);
    return NULL;
}
```

SriRamanujam commented on src/hyperv/hyperv_network_api_v1.c:
```diff
@@ -264,6 +264,81 @@ hyperv1ConnectNumOfNetworks(virConnectPtr conn)
+int
+hyperv1NetworkSetAutostart(virNetworkPtr network ATTRIBUTE_UNUSED,
+                           int autostart)
+{
```
Since this function is a no-op, why not always return -1?

Maners replied:
you can enable autostart but not disable - like in the esx driver.

SriRamanujam commented on src/hyperv/hyperv_network_api_v1.h:
```diff
@@ -32,6 +32,11 @@ int hyperv1ConnectListDefinedNetworks(virConnectPtr conn, char **const names,
         int maxnames);
 virNetworkPtr hyperv1NetworkLookupByName(virConnectPtr conn, const char *name);
 int hyperv1ConnectNumOfDefinedNetworks(virConnectPtr conn);
+char * hyperv1NetworkGetXMLDesc(virNetworkPtr network, unsigned int flags);
```
Should be `char *hyperv1NetworkGetXMLDesc`

SriRamanujam reviewed (changes requested):
I don't mean to be too nitpicky, but yeah. As usual, let me know if I missed anything or didn't pick up on some nuance.

Maners commented:
@SriRamanujam thanks for reviewing :-) I've addressed all the issues and force-pushed to keep clean commit history, except:

* the variable passed as const char *filter does not need cleanup as it's not heap allocated
* hostname is safe to use as it's fetch from Msvm_ComputerSystem by the calling code already so there's no need to lookup same info twice
* size_t for count variable should not be used because count is used as return value (assigned to "ret" right before cleanup label) of functions that declare int return type.
